qclib/gates/mcg.py

```diff
@@ -0,0 +1,276 @@
+            target_qubit=target,
+            mode="noancilla"
+        )
+    elif num_ctrl == 3:
+        self.append(C3XGate(), [*controls[:], target], [])
+    else:
+        for m in range(2):
+            for i, _ in enumerate(controls):      # action part
+                if i < num_ctrl - 2:
+                    if targets[i] != target or relative_phase:  # gate cancelling
+                        theta = np.pi / 4.
+
+                        # cancel rightmost gates of action part with leftmost gates of reset part
+                        if (relative_phase and targets[i] == target and m == 1):
+                            self.cx(ancillae[num_ancilla - i - 1], targets[i])
+                            self.u(theta=theta, phi=0., lam=0., qubit=targets[i])
+                            self.cx(controls[num_ctrl - i - 1], targets[i])
+                            self.u(theta=theta, phi=0., lam=0., qubit=targets[i])
+                        else:
+                            self.u(theta=-theta, phi=0., lam=0., qubit=targets[i])
+                            self.cx(controls[num_ctrl - i - 1], targets[i])
+                            self.u(theta=-theta, phi=0., lam=0., qubit=targets[i])
+                            self.cx(ancillae[num_ancilla - i - 1], targets[i])
+
+                    else:
+                        self.ccx(
+                            control_qubit1=controls[num_ctrl - i - 1],
+                            control_qubit2=ancillae[num_ancilla - i - 1],
+                            target_qubit=targets[i]
+                        )
+                else:
+                    theta = np.pi / 4.
+
+                    self.u(theta=-theta, phi=0., lam=0., qubit=targets[i])
+                    self.cx(controls[num_ctrl - i - 2], targets[i])
+                    self.u(theta=-theta, phi=0., lam=0., qubit=targets[i])
+                    self.cx(controls[num_ctrl - i - 1], targets[i])
+                    self.u(theta=theta, phi=0., lam=0., qubit=targets[i])
+                    self.cx(controls[num_ctrl - i - 2], targets[i])
+                    self.u(theta=theta, phi=0., lam=0., qubit=targets[i])
+
+                    break
+
+            for i, _ in enumerate(ancillae[1:]):      # reset part
+                theta = np.pi / 4.
+                self.cx(ancillae[i], ancillae[i + 1])
+                self.u(theta=theta, phi=0., lam=0., qubit=ancillae[i + 1])
+                self.cx(controls[2 + i], ancillae[i + 1])
+                self.u(theta=theta, phi=0., lam=0., qubit=ancillae[i + 1])
+
+
+            if action_only:
+                break
+
+    if ctrl_state is not None:
+        for i, ctrl in enumerate(ctrl_state[::-1]):
+            if ctrl == '0':
+                self.x(controls[i])
+
+
+QuantumCircuit.mcx_v_chain_dirty = mcx_v_chain_dirty
+
+
+def linear_depth_mcv(
+    self,
+    x,
+    z,
+    controls: Union[QuantumRegister, List[Qubit]],
+    target: Qubit,
+    ctrl_state: str=None
+):
+
```

Comment:
missing docstring

Reply:
We have to do a full style check before merge.